File: src/main/java/xyz/phanta/cbtweaker/hatch/HatchTileEntity.java
package xyz.phanta.cbtweaker.hatch;

import io.github.phantamanta44.libnine.capability.provider.CapabilityBroker;
import io.github.phantamanta44.libnine.tile.L9TileEntityTicking;
import io.github.phantamanta44.libnine.tile.RegisterTile;
import io.github.phantamanta44.libnine.util.data.serialization.AutoSerialize;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.common.capabilities.ICapabilityProvider;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.buffer.BufferGroup;
import xyz.phanta.cbtweaker.common.MachineDataHolder;
import xyz.phanta.cbtweaker.gui.inventory.UiElement;
import xyz.phanta.cbtweaker.multiblock.MultiBlockControllerTileEntity;
import xyz.phanta.cbtweaker.util.block.AutoExportHandler;

@RegisterTile(CbtMod.MOD_ID)
public class HatchTileEntity extends L9TileEntityTicking {

    @AutoSerialize
    private final MachineDataHolder<HatchData<?>> dataHolder = new MachineDataHolder<>(() -> {
        IBlockState state = getWorld().getBlockState(getPos());
        HatchBlock block = (HatchBlock)state.getBlock();
        return new HatchData<>(
                this, block.getHatchType(), state.getValue(block.getTierProperty()));
    });

    public HatchTileEntity() {
        markRequiresSync();
        setInitialized();
    }

    @Override
    protected ICapabilityProvider initCapabilities() {
        CapabilityBroker caps = new CapabilityBroker();
        dataHolder.getData().attachCapabilities(caps::with);
        return caps;
    }

    public HatchType<?> getHatchType() {
        return dataHolder.getData().getHatchType();
    }

    public int getTier() {
        return dataHolder.getData().getHatchTier();
    }

    public AutoExportHandler getExportHandler() {
        return dataHolder.getData().getExportHandler();
    }

    public UiElement createUiElement() {
        return dataHolder.getData().createUiElement();
    }

    @Override
    public void setDirty() {
        super.setDirty();
    }

    public void associate(MultiBlockControllerTileEntity mbCtrl) {
        dataHolder.getData().associate(mbCtrl);
    }

    public void disassociate(MultiBlockControllerTileEntity mbCtrl) {
        dataHolder.getData().disassociate(mbCtrl);
    }

    public void addToGroup(BufferGroup group) {
        dataHolder.getData().addToGroup(group);
    }

    public boolean handleInteraction(EntityPlayer player, EnumHand hand, EnumFacing face) {
        World world = getWorld();
        BlockPos pos = getPos();
        return dataHolder.getData().handleInteraction(world, pos, world.getBlockState(pos), player, hand, face);
    }

    public void dropContents() {
        dataHolder.getData().dropContents(getWorld(), getPos());
    }

    @Override
    protected void tick() {
        dataHolder.getData().tick();
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/hatch/HatchTileEntity.kt
package st.evening.mc.cbtweaker.hatch

import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.ITickable
import net.minecraft.util.math.BlockPos
import net.minecraftforge.common.capabilities.Capability
import st.evening.mc.cbtweaker.buffer.BufferGroup
import st.evening.mc.cbtweaker.common.LazyTileEntity
import st.evening.mc.cbtweaker.gui.inventory.UiElement
import st.evening.mc.cbtweaker.multiblock.MultiBlockControllerTileEntity
import st.evening.mc.cbtweaker.util.sync.TileEntitySyncProxy
import st.evening.mc.prelude.api.util.collection.CapabilityMap
import st.evening.mc.prelude.api.util.game.ServerSide

class HatchTileEntity : LazyTileEntity<HatchData<*>>(), ITickable {
    private val capabilities: CapabilityMap by lazy {
        CapabilityMap().also { data.attachCapabilities(it) }
    }

    override fun initData(): HatchData<*> {
        val state = world.getBlockState(getPos())
        val block = state.getBlock() as HatchBlock
        val data = HatchData(this, block.hatchType, state.getValue(block.tierProperty))
        data.getSyncState()?.let {
            initSync(TileEntitySyncProxy(this, it))
        }
        return data
    }

    val hatchType: HatchType<*>
        get() = data.hatchType

    val tier: Int
        get() = data.hatchTier

    val exportHandler: HatchData<*>.HatchAutoExportHandler?
        get() = data.exportHandler

    override fun hasCapability(capability: Capability<*>, facing: EnumFacing?): Boolean = capability in capabilities

    override fun <T : Any> getCapability(capability: Capability<T>, facing: EnumFacing?): T? = capabilities[capability]

    fun associate(mbCtrl: MultiBlockControllerTileEntity) {
        data.associate(mbCtrl)
    }

    fun disassociate(mbCtrl: MultiBlockControllerTileEntity) {
        data.disassociate(mbCtrl)
    }

    fun addToGroup(group: BufferGroup) {
        data.addToGroup(group)
    }

    override fun update() {
        data.tick()
    }

    fun handleInteraction(
        state: IBlockState, player: EntityPlayer, hand: EnumHand,
        face: EnumFacing, hitX: Float, hitY: Float, hitZ: Float
    ): Boolean = data.handleInteraction(state, player, hand, face, hitX, hitY, hitZ)

    @ServerSide
    fun handleBlockUpdate(state: IBlockState, fromBlock: Block, fromPos: BlockPos) {
        data.handleBlockUpdate(state, fromBlock, fromPos)
    }

    @ServerSide
    fun handleDestruction(state: IBlockState) {
        data.handleDestruction(state)
    }

    fun createUiElement(): UiElement? = data.createUiElement()
}
